src/metrics.jl

```diff
@@ -239,7 +239,7 @@ corr_dist(a::AbstractArray, b::AbstractArray) = evaluate(CorrDist(), a, b)
 result_type(::CorrDist, a::AbstractArray, b::AbstractArray) = result_type(CosineDist(), a, b)
 
 # ChiSqDist
-@inline eval_op(::ChiSqDist, ai, bi) = abs2(ai - bi) / (ai + bi)
+@inline eval_op(::ChiSqDist, ai, bi) = ifelse(ai != bi, abs2(ai - bi) / (ai + bi), zero(abs2(ai - bi) / (ai + bi)))
```

Comment:
If you use the same expression twice, better store the result in a variable and call `zero` on it when `ai != bi`.

Reply:
The 2nd occurrence is only for type inference, it's not calculated (checked on v0.6.1).
I've tried your suggestion: `@code_native` shows that in this particular case the compiler was smart to reorder the operations and eliminate the temporary var, so nothing is put to the stack.
But I would rather not tempt the compiler in performance critical methods like this.

Reply:
My suggestion was about readability. AFAICT it could only be faster than repeating the expression, and never slower.

Reply:
I've implemented your suggestion, because in this case the compiler still generates the optimal code.
I'm just not sure we both agreed that the 2nd occurrence of the formula is eliminated during type inference, so the current version (`d=...; ifelse(ai != bi, d, 0)`) have suboptimal evaluation order, and we rely on LLVM to fix it.